modules/tensor_mechanics/src/postprocessors/InteractionIntegral.C

```diff
@@ -0,0 +1,346 @@
+  }
+
+  for (_qp = 0; _qp < _qrule->n_points(); _qp++)
+    sum += _JxW[_qp] * _coord[_qp] * computeQpIntegral();
+  return sum;
+}
+
+void
+InteractionIntegral::computeAuxFields(const MooseEnum sif_mode,
+                                      RankTwoTensor & aux_stress,
+                                      RankTwoTensor & grad_disp)
+{
+  RealVectorValue k(0);
+  if (sif_mode == "KI")
+    k(0) = 1;
+
+  else if (sif_mode == "KII")
+    k(1) = 1;
+
+  else if (sif_mode == "KIII")
+    k(2) = 1;
+
+  Real t = _theta;
+  Real t2 = _theta / 2;
+  Real tt2 = 3 * _theta / 2;
+  Real st = std::sin(t);
+  Real ct = std::cos(t);
+  Real st2 = std::sin(t2);
+  Real ct2 = std::cos(t2);
+  Real stt2 = std::sin(tt2);
+  Real ctt2 = std::cos(tt2);
+  Real ct2sq = std::pow(ct2, 2);
+  Real ct2cu = std::pow(ct2, 3);
+  Real sqrt2PiR = std::sqrt(2 * libMesh::pi * _r);
+
+  // Calculate auxiliary stress tensor
+  Real s11 = 1 / sqrt2PiR * (k(0) * ct2 * (1 - st2 * stt2) - k(1) * st2 * (2 + ct2 * ctt2));
+  Real s22 = 1 / sqrt2PiR * (k(0) * ct2 * (1 + st2 * stt2) + k(1) * st2 * ct2 * ctt2);
+  Real s12 = 1 / sqrt2PiR * (k(0) * ct2 * st2 * ctt2 + k(1) * ct2 * (1 - st2 * stt2));
+  Real s13 = -1 / sqrt2PiR * k(2) * st2;
+  Real s23 = 1 / sqrt2PiR * k(2) * ct2;
+  // plain stress
+  // Real s33 = 0;
+  // plain strain
+  Real s33 = _poissons_ratio * (s11 + s22);
+
+  aux_stress(0, 0) = s11;
+  aux_stress(0, 1) = s12;
+  aux_stress(0, 2) = s13;
+  aux_stress(1, 0) = s12;
+  aux_stress(1, 1) = s22;
+  aux_stress(1, 2) = s23;
+  aux_stress(2, 0) = s13;
+  aux_stress(2, 1) = s23;
+  aux_stress(2, 2) = s33;
+
+  // Calculate x1 derivative of auxiliary displacements
+  Real du11 = k(0) / (4 * _shear_modulus * sqrt2PiR) *
+                  (ct * ct2 * _kappa + ct * ct2 - 2 * ct * ct2cu + st * st2 * _kappa + st * st2 -
+                   6 * st * st2 * ct2sq) +
+              k(1) / (4 * _shear_modulus * sqrt2PiR) *
+                  (ct * st2 * _kappa + ct * st2 + 2 * ct * st2 * ct2sq - st * ct2 * _kappa +
+                   3 * st * ct2 - 6 * st * ct2cu);
+
+  Real du21 = k(0) / (4 * _shear_modulus * sqrt2PiR) *
+                  (ct * st2 * _kappa + ct * st2 - 2 * ct * st2 * ct2sq - st * ct2 * _kappa -
+                   5 * st * ct2 + 6 * st * ct2cu) +
+              k(1) / (4 * _shear_modulus * sqrt2PiR) *
+                  (-ct * ct2 * _kappa + 3 * ct * ct2 - 2 * ct * ct2cu - st * st2 * _kappa +
+                   3 * st * st2 - 6 * st * st2 * ct2sq);
+
+  Real du31 = k(2) / (_shear_modulus * sqrt2PiR) * (st2 * ct - ct2 * st);
+
+  grad_disp(0, 0) = du11;
+  grad_disp(0, 1) = du21;
+  grad_disp(0, 2) = du31;
+}
+
+void
+InteractionIntegral::computeTFields(RankTwoTensor & aux_stress, RankTwoTensor & grad_disp)
+{
+
+  Real t = _theta;
+  Real st = std::sin(t);
+  Real ct = std::cos(t);
+  Real stsq = std::pow(st, 2);
+  Real ctsq = std::pow(ct, 2);
+  Real ctcu = std::pow(ct, 3);
+  Real oneOverPiR = 1.0 / (libMesh::pi * _r);
+
```

Comment:
Call
`aux_stress.zero();`
first, then we can eliminate the zeroing out of individual terms.

Reply:
Actually, I guess it's not really necessary to do that because it's created just before this is called, but maybe it makes the code more readable and safer. The same is true for `grad_disp`, both in this function and in `computeAuxFields()`.